### ISSUE TITLE
IsEqualTo  wrong check

### ISSUE THREAD

**Issue by timur-sokolov** [tpierrain/NFluent#335]
**Bug Type**
Please pick one:
- [x] a check failed to detect an error (false negative), i.e. a test is green while it should be red.
- [ ] a check raised un existing error (false positive), i.e a test is red while it should be green.
- [ ] an error message is invalid/incomplete (please provide samples)
- [ ] a ran into an exception.
- [ ] other.

**Describe the bug**
IsEqualTo is green while it should be red. See to reproduce section

**To Reproduce**
```
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NFluent;

namespace NFluentIssue
{
    [TestClass]
    public class NFluentTest
    {
        [TestMethod]
        public void IsEqualToIssue()
        {
            var input = new { data = "testData" };
            var json = JsonConvert.SerializeObject(input);
            
            var obj = JObject.Parse(json);

            Check.That(obj["data"]).IsEqualTo("Why did it pass?");
        }

        [TestMethod]
        public void IsEqualToIssue_WithToString()
        {
            var input = new { data = "testData" };
            var json = JsonConvert.SerializeObject(input);

            var obj = JObject.Parse(json);

            Check.That(obj["data"].ToString()).IsEqualTo("testData");
        }
    }    
}
```

**Expected behavior**
IsEqualToIssue test method should fail, but both tests passed.

**Screenshots**
See attached screenshot.
![image](https://user-images.githubusercontent.com/34766753/152319034-70ac7288-e5e5-41dc-a8c7-478eec83087d.png)

**Desktop (please complete the following information):**
- NFluent version: 2.7.2
- Net Version: .NET 6
- IDE: VS 2022 [eg VS 2019, VSCode, Rider...]

**Comment by dupdob** [comment]
Thanks for reporting this. I will look into this in the coming hours.
A fix will be delivered ASAP

**Comment by dupdob** [comment]
## Analysis
I have reproduced this issue and analyzed it. NewtonSoft `JObject` confuses NFluent equality logic for two reasons:
1. There is an implicit conversion from `string` to `JObject` 
2. `JObject` exposes an enumeration interface (`Enumerable`)

For context, NFluent implements both a typed and an untyped version of `IsEqualTo` (i.e `IsEqualTo(object expected)` and `IsEqualTo(T expected)` where `T` is the type of the _sut_ (i.e. the checked object/value). Here, due to the implicit conversion, `"Why did it pass?"` is converted to a `JObject` instance.
Then (2) kicks in, NFluent uses the enumeration comparison to compare the _sut_ with the expected value and both appears as empty enumeration, so they are declared as equal.

## Workaround
You can use `ToString` (as you did in your second test) to secure the result. You can also force the type of the expected (to object, as in `Check.That(obj["data"].IsEqualTo("Why dit it pass?");`, which will prevent enumeration comparison to happen and result in a failing check.

## Fix
At this stage, I am not sure what the proper behavior is. The next major version of NFluent would help here, as it allows to specify equality comparison logic per type.

@timur-sokolov, which behavior do you expect:
1. `Check.That(obj["data"]).IsEqualTo("Why did it pass?");` must fail
2. `Check.That(obj["data"]).IsEqualTo("testData");` must pass

Both requires some form of 'NewtonSoft' specific hack. A cleaner fix may be possible for 1. but it would trigger behavior change for other type as well. I am not sure I can achieve 2, because `obj["data"] == new JObject("testData")` is false, so not sure how I can flag them as identical.

**Comment by timur-sokolov** [comment]
What about failing in both cases with a clear message that comparing values has different types? In that case, the user will do something to cast values to the same type.

**Comment by dupdob** [comment]
Ok, that's option 1 then (I should have been specific in my proposals).
I can fix this by removing the typed overload for `IsEqualTo`. But I would like to avoid it because it would degrade autocompletion (not a big deal) but mostly this would be an accidental fix.
For example, this code would still work:
`Check.That(obj["data"]).IsEqualTo(new JObject());`

**Comment by dupdob** [comment]
I have what looks like a good fix. But,  I still have a problem with the error message. `JValue` is reported as an (empty) enumeration so this does not look great.
I will keep on improving.
If you need a fix for that quickly, I can push a pre release version that will provide the fix, albeit with an unclear error message

**Comment by dupdob** [comment]
@timur-sokolov : I just published a pre release version with the fix (V2.8 beta 0248).
Could you please test it and confirm it fixes your problem please ?
If it is ok, I will promote to stable

**Comment by timur-sokolov** [comment]
@dupdob It works. Thanks

**Comment by MendelMonteiro** [comment]
@dupdob , commenting out the typed `IsEqualTo()` method is a breaking change.  Any code that called `IsEqualTo(default)` will break where `T` is not a reference type.

**Comment by dupdob** [comment]
thanks, I failed to anticipate that impact. I will document it.
## Analysis
This is an impossible conundrum to solve: one signature break `IsEqualTo(default)`, leading to false positive, the other one leads to false negative with badly designed classes (i.e. not truly OO).
 **Dealing with false negative** has priority over false positive, as people can correct their assertions when they detect the error. 

## Workaround
One can use `Is(..)` as a replacement to `IsEqualTo`. `Is` is only available in strongly typed form, so `Check.That(...).Is(default)` works as expected.
I plan to add a rule in `NFluent.Analyzer` to detect `IsEqualTo(default)` patterns and  offer the correct alternative.

Also, I can add a `IsDefault` check looks even better to me.

**Comment by dupdob** [comment]
Well, this is awkward: `Is` and `IsDefaultValue` are already available for at least a couple of release. But this has not been documented.
sorry for that

**Comment by dupdob** [comment]
here is a full update on this:
1. `IsDefault` exists since V2.3.
1. `Is` has been released as part of V2.8
1. Correction, there is no false negative due to the typed version of `IsEqualTo`, my bad. The false negative was due to a combination of this signature and the internal NFluent logic, and adjusting internal logic removed the risk. But the `IsEqualTo` typed signature leads to surprising error messages.

**Comment by MendelMonteiro** [comment]
It seems that `Is` is not available in 2.8.  Intellisense can't find it and when I check the code I see it's not present in the `ObjectCheckExtensions.cs` file at the 2.8 tag revision:

https://github.com/tpierrain/NFluent/blob/9c463d9a0d81e15fab9d3eeb61a2d17ce0e89e1f/code/src/NFluent/Checks/ObjectCheckExtensions.cs

But it is present on `master`: https://github.com/tpierrain/NFluent/blob/72d78e576f618054f034fa410154f68996822374/code/src/NFluent/Checks/ObjectCheckExtensions.cs#L96

**Comment by dupdob** [comment]
My bad. `Is` has not been released. I did make some tests BEFORE saying this was available (and the tests were successful). Not sure what I did wrong, but repeating those tests now clearly shows that `Is` is NOT in V2.8

sorry for the mix up
